docs/development/typescript.md

```diff
@@ -0,0 +1,56 @@
+  - [Caveats](#caveats)
+  - [Prerequisitions](#prerequisitions)
+  - [Implementation](#implementation)
+
+## Project references
+Kibana has crossed the 1.5m LoC mark. The current situation creates some scaling problems when the default out-of-the-box setup stops working. As a result, developers suffer from slow project compilation and IDE unresponsiveness. As a part of [Developer Experience project](https://github.com/elastic/kibana/projects/63), we are migrating our tooling to use built-in TypeScript features addressing the scaling problems - [project references](https://www.typescriptlang.org/docs/handbook/project-references.html) & [incremental builds](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-4.html#faster-subsequent-builds-with-the---incremental-flag).
+
+In a nutshell - instead of compiling the whole Kibana codebase at once, this setup enforces splitting the code base into independent projects that form a directed acyclic graph (DAG). This allows the TypeScript compiler (`tsc`) to apply several advanced optimizations:
+- Every project emits *public* interfaces in the form of `d.ts` type declarations generated by the TypeScript compiler
+- These generated `d.ts` type declarations are used whenever a referenced project is imported in a depending project
+- This makes it possible to determine which project needs rebuilding when the source code has changed to use a more aggressive caching strategy.
+More details are available in the [official docs](https://www.typescriptlang.org/docs/handbook/project-references.html)
+
+### Caveats
+This architecture imposes several limitations to which we must comply:
+- Projects cannot have circular dependencies. Even though the Kibana platform doesn't support circular dependencies between Kibana plugins, TypeScript (and ES6 modules) does allow circular imports between files. So in theory, you may face a problem when migrating to the TS project references and you will have to resolve this circular dependency.
+- A project must emit its type declaration. It's not always possible to generate a type declaration if the compiler cannot infer a type. There are two basic cases:
+1. Your plugin exports a type inferring an internal type declared in Kibana codebase. In this case, you'll have to either export an internal type or to declare an exported type explicitly.
+2. Your plugin exports something inferring a type from a 3rd party library that doesn't export this type. To fix the problem, you have to declare the exported type manually.
+
+### Prerequisitions
+Since `tsc` doesn't support circular project references, the migration order does matter. You can migrate your plugin only when all the plugin dependencies already have migrated. It creates a situation where commonly used plugins (such as `data` or `kibana_react`) have to migrate first.
+
+### Implementation
+- make sure all the plugins listed as dependencies in `kibana.json` file have migrated to TS project references.
```

Comment:
nit: I think a numbered list makes more sense here since you will generally start at the top and work down

Reply:
I had to revert it in https://github.com/elastic/kibana/pull/78586/commits/6efc968f8595599d5560d7580f9de812137b7ab9 to the unordered list. asciidoctor fails validation with
```
18:06:10 INFO:build_docs:asciidoctor: WARNING: developer/best-practices/typescript.asciidoc: line 35: list item index: expected 2, got 1
18:06:10 INFO:build_docs:asciidoctor: WARNING: developer/best-practices/typescript.asciidoc: line 59: list item index: expected 2, got 1
18:06:10 INFO:build_docs:asciidoctor: WARNING: developer/best-practices/typescript.asciidoc: line 60: list item index: expected 3, got 1
18:06:10 INFO:build_docs:asciidoctor: WARNING: developer/best-practices/typescript.asciidoc: line 61: list item index: expected 4, got 1
18:06:10 INFO:build_docs:asciidoctor: WARNING: developer/best-practices/typescript.asciidoc: line 62: list item index: expected 5, got 1
```
I tried `1., 2., etc.` style as well, but the numeration started from `1` after `source` block. 😞